File: app/src/main/java/dev/favourdevlabs/cleanthes/security/CryptoManager.java
package dev.favourdevlabs.cleanthes.security;

import android.util.Base64;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import java.security.SecureRandom;

public class CryptoManager {

    private static final String TRANSFORMATION = "AES/GCM/NoPadding";

    private static final int GCM_TAG_LENGTH_BITS = 128;

    private static final int IV_LENGTH_BYTES = 12;

    public static String encrypt(String plaintext, SecretKey secretKey) throws Exception {

        byte[] iv = generateIV();

        Cipher cipher = Cipher.getInstance(TRANSFORMATION);
        GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, parameterSpec);

        byte[] ciphertext = cipher.doFinal(plaintext.getBytes("UTF-8"));

        byte[] ivPlusCiphertext = new byte[IV_LENGTH_BYTES + ciphertext.length];
        System.arraycopy(iv, 0, ivPlusCiphertext, 0, IV_LENGTH_BYTES);
        System.arraycopy(ciphertext, 0, ivPlusCiphertext, IV_LENGTH_BYTES, ciphertext.length);

        return Base64.encodeToString(ivPlusCiphertext, Base64.NO_WRAP);
    }

    public static String decrypt(String encryptedData, SecretKey secretKey) throws Exception {

        byte[] ivPlusCiphertext = Base64.decode(encryptedData, Base64.NO_WRAP);

        byte[] iv = new byte[IV_LENGTH_BYTES];
        byte[] ciphertext = new byte[ivPlusCiphertext.length - IV_LENGTH_BYTES];
        System.arraycopy(ivPlusCiphertext, 0, iv, 0, IV_LENGTH_BYTES);
        System.arraycopy(ivPlusCiphertext, IV_LENGTH_BYTES, ciphertext, 0, ciphertext.length);

        Cipher cipher = Cipher.getInstance(TRANSFORMATION);
        GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, parameterSpec);

        byte[] plaintext = cipher.doFinal(ciphertext);

        return new String(plaintext, "UTF-8");
    }

    private static byte[] generateIV() {
        byte[] iv = new byte[IV_LENGTH_BYTES];
        new SecureRandom().nextBytes(iv);
        return iv;
    }
}

File: app/src/main/java/dev/favourdevlabs/cleanthes/security/CryptoManager.kt
package dev.favourdevlabs.cleanthes.security

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

object CryptoManager {

    private const val TRANSFORMATION      = "AES/GCM/NoPadding"
    private const val GCM_TAG_LENGTH_BITS = 128
    private const val IV_LENGTH_BYTES     = 12

    @Throws(Exception::class)
    fun encrypt(plaintext: String, secretKey: SecretKey): String {
        val iv = generateIV()
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))
        val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

        val ivPlusCiphertext = ByteArray(IV_LENGTH_BYTES + ciphertext.size)
        System.arraycopy(iv, 0, ivPlusCiphertext, 0, IV_LENGTH_BYTES)
        System.arraycopy(ciphertext, 0, ivPlusCiphertext, IV_LENGTH_BYTES, ciphertext.size)

        return Base64.encodeToString(ivPlusCiphertext, Base64.NO_WRAP)
    }

    @Throws(Exception::class)
    fun decrypt(encryptedData: String, secretKey: SecretKey): String {
        val ivPlusCiphertext = Base64.decode(encryptedData, Base64.NO_WRAP)

        val iv         = ivPlusCiphertext.copyOfRange(0, IV_LENGTH_BYTES)
        val ciphertext = ivPlusCiphertext.copyOfRange(IV_LENGTH_BYTES, ivPlusCiphertext.size)

        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, secretKey, GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv))

        return String(cipher.doFinal(ciphertext), Charsets.UTF_8)
    }

    private fun generateIV(): ByteArray =
        ByteArray(IV_LENGTH_BYTES).also { SecureRandom().nextBytes(it) }
}
